### PR TITLE
Implement add/update support into android logins-api

### PR DESCRIPTION
**Pull request by thomcc**
Not required for lockbox short term, but necessary long term. This also implements it in the in-memory version.

I also clarified some of the documentation slightly.

### CODE REVIEW

File: logins-api/android/library/src/main/java/org/mozilla/sync15/logins/LoginsStorage.kt [hunk 1]
```diff
@@ -59,4 +59,35 @@ interface LoginsStorage : Closeable {
...
+    /**
+     * Insert the provided login into the database.
+     *
+     * This function ignores values in metadata fields (`timesUsed`,
+     * `timeCreated`, `timeLastUsed`, and `timePasswordChanged`).
+     *
+     * A GUID may be provided, however if `login` has an empty
+     * GUID then one will be generated automatically.
+     *
+     * This will return an error result if a guid is provided but
```
Comment (ncalexan) [inline logins-api/android/library/src/main/java/org/mozilla/sync15/logins/LoginsStorage.kt]: nit: capital "GUID" (for consistency with the above).

File: logins-api/android/library/src/main/java/org/mozilla/sync15/logins/LoginsStorage.kt [hunk 2]
```diff
@@ -59,4 +59,35 @@ interface LoginsStorage : Closeable {
...
      */
     fun list(): SyncResult<List<ServerPassword>>
+
+    /**
+     * Insert the provided login into the database.
+     *
+     * This function ignores values in metadata fields (`timesUsed`,
+     * `timeCreated`, `timeLastUsed`, and `timePasswordChanged`).
+     *
+     * A GUID may be provided, however if `login` has an empty
```
Comment (ncalexan) [inline logins-api/android/library/src/main/java/org/mozilla/sync15/logins/LoginsStorage.kt]: uber-nit: This is just a little awkward.  Maybe just: "If `login` does not have a `id` field, or has an empty `id` field, then a GUID will be generated automatically."  What style of GUID do we generate?  (8 chars or "{guid}" style?)

File: logins-api/android/library/src/main/java/org/mozilla/sync15/logins/LoginsStorageException.kt [hunk 3]
```diff
@@ -8,13 +8,35 @@
...
 class SyncAuthInvalidException(msg: String): LoginsStorageException(msg)
 
-// This doesn't really belong in this file...
+/**
+ * Thrown if `lock()`/`unlock()` pairs don't match up.
+ */
 class MismatchedLockException(msg: String): LoginsStorageException(msg)
+
+/**
+ * Thrown if `update()` is performed with a record whose ID does not exist
```
Comment (ncalexan) [inline logins-api/android/library/src/main/java/org/mozilla/sync15/logins/LoginsStorageException.kt]: nit: full sentence.

File: logins-api/android/library/src/main/java/org/mozilla/sync15/logins/MemoryLoginsStorage.kt [hunk 4]
```diff
@@ -149,6 +194,27 @@ class MemoryLoginsStorage(private var list: List<ServerPassword>) : Closeable, L
...
     }
 
+    private fun checkValid(login: ServerPassword) {
+        if (login.hostname == "") {
+            throw InvalidRecordException("Invalid login: Hostname is empty")
+        }
+        if (login.password == "") {
+            throw InvalidRecordException("Invalid login: Password is empty")
+        }
+        if (login.password == "") {
```
Comment (ncalexan) [inline logins-api/android/library/src/main/java/org/mozilla/sync15/logins/MemoryLoginsStorage.kt]: Copy-pasta error?  Duplicated.

Comment (ncalexan) [reply logins-api/android/library/src/main/java/org/mozilla/sync15/logins/MemoryLoginsStorage.kt]: Maybe you wanted to null check?

File: logins-sql/ffi/src/error.rs [hunk 5]
```diff
@@ -126,6 +126,16 @@ pub enum ExternErrorCode {
     /// Indicates the FxA credentials are invalid, and should be refreshed.
     AuthInvalidError = 1,
 
+    /// Returned from an `update()` call where the record ID did not exist.
+    NoSuchRecord = 2,
+
+    /// Returned from an `add()` call that was provided an ID, where the ID
+    /// Already existed.
```
Comment (ncalexan) [inline logins-sql/ffi/src/error.rs]: nit: downcase A.

File: logins-api/android/library/src/main/java/org/mozilla/sync15/logins/LoginsStorage.kt (hunk 2 above)
Comment (thomcc) [reply logins-api/android/library/src/main/java/org/mozilla/sync15/logins/LoginsStorage.kt]: 12 base64url encoded random characters.

File: logins-api/android/library/src/main/java/org/mozilla/sync15/logins/MemoryLoginsStorage.kt (hunk 4 above)
Comment (thomcc) [reply logins-api/android/library/src/main/java/org/mozilla/sync15/logins/MemoryLoginsStorage.kt]: Accidentally copypasted it. Password can't be null.

**Pull request merged**